Code/Framework/AzCore/AzCore/Statistics/StatisticalProfilerProxy.h

```diff
@@ -7,28 +7,24 @@
  */
 #pragma once
 
-#include <AzCore/std/chrono/types.h>
-#include <AzCore/std/parallel/shared_spin_mutex.h>
-#include <AzCore/std/parallel/scoped_lock.h>
-#include <AzCore/std/containers/bitset.h>
-#include <AzCore/std/string/string.h>
 #include <AzCore/Interface/Interface.h>
 #include <AzCore/Statistics/StatisticalProfiler.h>
-#include <AzCore/Debug/Profiler.h>
-
+#include <AzCore/std/containers/unordered_map.h>
+#include <AzCore/std/parallel/shared_spin_mutex.h>
 
 #if defined(AZ_STATISTICAL_PROFILING_ENABLED)
 
 #if defined(AZ_PROFILE_SCOPE)
 #undef AZ_PROFILE_SCOPE
 #endif // #if defined(AZ_PROFILE_SCOPE)
 
-#define AZ_PROFILE_SCOPE(profiler, scopeNameId) \
+#define AZ_PROFILE_SCOPE(budget, scopeNameId, ...) \
     static const AZStd::string AZ_JOIN(blockName, __LINE__)(scopeNameId); \
-    AZ::Statistics::StatisticalProfilerProxy::TimedScope AZ_JOIN(scope, __LINE__)(profiler, AZ_JOIN(blockName, __LINE__));
+    AZ::Statistics::StatisticalProfilerProxy::TimedScope AZ_JOIN(scope, __LINE__)(AZ_CRC_CE(#budget), AZ_JOIN(blockName, __LINE__));
```

Comment:
This code is dangerous.

The static const should really be an `AZStd::fixed_string`
Perhaps an `AZStd::fixed_string<64>` or `AZStd::fixed_string<128>`

There is no guarantee where an `AZ_PROFILE_SCOPE` would be used at and if it is used when the AZ allocators are not available, this macro could cause a crash if the `scopeNameId` happens to be longer than the short string buffer of 16 bytes.

Reply:
Maybe it's better to remove "static". Simple "const AZStd::string" should suffice.

Reply:
I'll certainly fix this

Reply:
Even only removing `static` could have issues if the `AZ_PROFILE_SCOPE` scope macro is used before the AZ Allocators are online.

The ComponentApplication constructor creates the AZ OSAllocator and AZ SystemAllocator if they are not created, so in most applications it will be on line quite early, but it is not guaranteed.

Reply:
Sure but in that case this particular `AZStd::string` wouldn't be the only memory issue occurring.  The statistical profiler types are reliant on the `SystemAllocator` being initialized before use.  Even the default `AZ_PROFILE_SCOPE` implementation requires those allocators, so we would need to rework how that macro is designed to support early use.  Dropping the `static` should be sufficient for now.

Reply:
But there is nothing implicit just using the AZ_PROFILE_SCOPE macro that would cause a crash if the profiler proxy isn't registered with an AZ::Interface.
The `StatisticalProfilerProxy::TimedScope` wouldn't hit the crash if the AZ SystemAllocator wasn't up in running, because the StatisticalProfilerProxy would not have an instance registered with an AZ::Interface making the code at least inert in the situation where it is used before the SystemAllocator is up and running.

`AZStd::string` doesn't such a protection and would allocator using the AZStd::allocator if it is initialized with a large enough string.

If anything to be safe we should either use a `std::string` if we need dynamic memory that isn't reliant on our AZ Allocator, or a `AZStd::fixed_string<256>` or `AZStd::basic_string<char, AZStd::char_traits<char>, static_buffer_allocator<256,1>` if we don't need dynamic memory and can allow a restriction on the block name string size.

Reply:
Even the `AZ::Interface` query requires at the very least the OS allocator to be initialized, does it not?  And as you mentioned, neither allocator is guaranteed until after the `ComponentApplication` is constructed so we are walking into questionable early macro usage territory here.  I'll completely agree that those "incorrect" calls should not cause a crash and also not advocating the dynamic memory string is actually necessary here, just pointing out it isn't the only issue in those cases.

Reply:
There are more issues here. The fact there's a `static` implicitly means there's a branch and the fact that every capture is going to potentially add an allocation for the string is going to further skew profiling. It seems possible though to update `StatisticalProfilerProxy.h` to use a `const char*` instead at quick glance. That does leave the mutex which is going to result in potentially high contention for something like a profiler. The spin in the mutex is just going to shift that time but doesn't resolve the contention issue. A better solution for tracking profiling information is to do something similar as the Event Logger and have thread local circular buffers that are daisy chained.

Reply:
Changing the `StatisticalProfilerProxy::StatIdType` type alias to `const char*` technically works but could be problematic as it gets passed down the chain and used as a key in the `StatisticsManager`'s internal map.  Then changing `StatisticalProfilerProxy::TimedScope` constructor is only delaying that inevitable allocation and makes the API a bit odd.  Ended up changing the type alias to `AZ::crc32` as keeping it a string seemed more trouble than it's worth for little to no benefit.  The stat entry still has to be registered before hand which provides an option to name it anyway.

As for the mutex issue, that kind of change is a bit out of scope for this review.  It certainly will need to be investigated.  Another local change that should lessen the impact (at least for now) is instead of conditionally redefining the `AZ_PROFILE_SCOPE` macro, which has hundreds of calls already, is to have this system utilize the `AZ_PROFILE_INTERVAL_SCOPED` macro instead.

Reply:
A Crc32 is good with me, for the TimedScope BlockName.